### PR TITLE
[WIP] Rework cost structs, csv generation

### PR DESCRIPTION
**Pull request by nicmr**
### Related issues:

#49 #50 

### Completed:
- Complete rework of structs in `package costs`.
Structs now have a more clearly defined purpose.
- `package csv` has been entirely reworked, renamed to `package invoicegen` and offers a better structured interface.
- [x] Margin and total showing up as 0
- [x] AWS costs that can not be associated with a ProjectID should be marked in a easily recognisable way, currently ProjectID field is simply empty.
- [x] Provide constructor functions, so structs won't have to be created manually
- [x] Add tests for new costcalc implementation (using a `testing` provider that mocks the apicall)
- [x] better dir hierarchy in s3
- [x] Add tests for new csv implementation

### CODE REVIEW

File: pkg/costs/aws.go [hunk 1]
```diff
@@ -94,9 +93,9 @@ func costsMonthlyAWS(month time.Time) (APICallResult, error) {
 	}
 
 	result := APICallResult{
-		Timestamp:  time.Now(),
-		Response:   output.String(),
-		CsvEntries: csvEntries,
+		Timestamp:      time.Now(),
```
Comment (cheld) [inline pkg/costs/aws.go]: Structure much better. I like the names

File: pkg/costs/costs.go [hunk 2]
```diff
@@ -11,20 +11,43 @@ import (
...
+	Month         string `csv:"Month" order:"0"`
+	ProjectID     string `csv:"ProjectID" order:"1"`
+	ContactPerson string `csv:"ContactPerson" order:"2"`
+	Amount        string `csv:"Amount" order:"3"`
+	Margin        string `csv:"Margin" order:"4"`
+	Total         string `csv:"Total" order:"5"`
 }
 
 // ToCsvString returns a csv-conformant string representation of apiResult as a
 func (apiResult APICallResult) ToCsvString() string {
```
Comment (cheld) [inline pkg/costs/costs.go]: The costs package should not have a dependency to CSV

File: pkg/costs/costs.go [hunk 3]
```diff
@@ -11,20 +11,43 @@ import (
...
+	Entries []Entry
+}
+
+// Entry is a struct that holds the relevant returned information returned by an APICall
+type Entry struct {
+	Month         string `csv:"Month" order:"0"`
+	ProjectID     string `csv:"ProjectID" order:"1"`
+	ContactPerson string `csv:"ContactPerson" order:"2"`
+	Amount        string `csv:"Amount" order:"3"`
+	Margin        string `csv:"Margin" order:"4"`
```
Comment (cheld) [inline pkg/costs/costs.go]: Entry is part of APICallResult. Margin is part of Entry. However, Margin is not part of APICallResult. 

You are mixing dependencies.

BTW, I just guess, but part of the problem might be that the structs do not have clear intention revealing names.

Comment (cheld) [reply pkg/costs/costs.go]: In other words, is APICallResult representing a call to the provider or representing the result of the cost calculation?

File: pkg/csv/csv.go [hunk 4]
```diff
@@ -5,59 +5,91 @@ package csv
...
+
+			// Approach 2:
+			// stringerType := reflect.TypeOf((*fmt.Stringer)(nil)).Elem()
+			// if !field.Type.Implements(stringerType) {
+			// 	return "", errors.New("Provided field does not fulfill fmt.Stringer interface: " + field.Name)
+			// }
+
+			// ensure required field tags are present
+			csvtag := field.Tag.Get("csv")
+			if csvtag == "" {
```
Comment (cheld) [inline pkg/csv/csv.go]: hmm...I don't under this line? Under which circumstance will this be true?

Comment (nicmr) [reply pkg/csv/csv.go]: Afaik it's a safe function, i.e. it shouldn't return errors or panic - if a tag is not present, it will return an empty string.

**Pull request merged**